modules/runtime-common/src/main/scala/com/snowplowanalytics/snowplow/runtime/AppHealth.scala

```diff
@@ -0,0 +1,204 @@
+ *
+ * {{{
+ * trait MyAppAlert
+ * case object BadPassword extends MyAppAlert
+ * case class MissingPermissions(moreDetails: String) extends MyAppAlert
+ * }}}
+ *
+ * Implement a `cats.Show` for the runtime services and for the alerts
+ *
+ * {{{
+ * val runtimeShow = Show[MyAppRuntimeService] {
+ *   case BadRowsOutputStream => "Bad rows output stream"
+ *   case EventsInputStream => "Events input stream"
+ * }
+ *
+ * val alertShow = Show[MyAppAlert] {
+ *   case BadPassword => "Bad password"
+ *   case MissingPermissions(moreDetails) => "Missing permissions " + moreDetails
+ * }
+ * }}}
+ *
+ * Implement a `cats.Eq` for the alerts, so we can alert on anything uniquely new
+ * {{{
+ * val alertEq = Eq.fromUniversalEquals[Alert] // Do it better than this
+ * }}}
+ *
+ * ==Environment initialization==
+ *
+ * Initialize the AppHealth as part of App Environment initialization:
+ *
+ * {{{
+ * appHealth <- AppHealth.init[F, MyAppAlert, MyAppRuntimeService]
+ * }}}
+ *
+ * Initialize a health probe, so the app reports itself as unhealthy whenever a required service
+ * becomes unhealthy
+ *
+ * {{{
+ * _ <- HealthProbe.resource(port, appHealthy)
+ * }}}
+ *
+ * Initialize the webhook, so the app reports new setup errors or whenever the setup configuration
+ * becomes healthy
+ *
+ * {{{
+ * _ <- Webhook.resource(config, appInfo, httpClient, appHealth)
+ * }}}
+ *
+ * And finally, register any runtime service that provides its own health reporter. Not all services
+ * fall into this category, but the source stream does fall into this category.
+ *
+ * {{{
+ * _ <- appHealth.addRuntimeHealthReporter(EventsInputStream, sourceAndAckIsHealthy)
+ * }}}
+ *
+ * ==Application processing==
+ *
+ * Once the application enters its processing phase, you can set the health status for any runtime
+ * service or setup configuration
+ *
+ * {{{
+ * // After catching an exception in the bad events stream
+ * _ <- Logger[F].error(e)("Problem with bad rows output stream")
+ * _ <- appHealth.becomeUnhealthyForRuntimeService(BadRowsOutputStream)
+ *
+ * // After bad rows stream becomes healthy again
+ * _ <- Logger[F].debug("Bad rows output stream is ok")
+ * _ <- appHealth.becomeHealthyForRuntimeService(BadRowsOutputStream)
+ *
+ * // After catching an exception with the external setup configuration
+ * // Note this will send an alert webhook
+ * _ <- Logger[F].error(e)("Problem with the provided password")
+ * _ <- appHealth.becomeUnhealthyForSetup(BadPassword)
+ *
+ * // After successful connection to the externally configured services
+ * // Note this will send the first hearbeat webhook
+ * _ <- Logger[F].error(e)("Everything ok with the provided setup configuration")
+ * _ <- appHealth.becomeHealthyForSetup
+ * }}}
+ *
+ * The application processing code does not need to explicitly send any monitoring alert or adjust
+ * the health probe return code.
+ */
+class AppHealth[F[_]: Monad, SetupAlert, RuntimeService] private (
```

Comment:
Looking at these type params and I'm wondering...if we could simply use:

```scala
final case class Service(name: String)
final case class Alert(message: String)
```
instead of type params? Just use them directly here in API. To convert from app specific stuff:

```scala
object Service {
  def from[APP_SERVICE: Show](service: APP_SERVICE) = Service(service.show)
}
```

I know, I know, it's not perfect, but maybe that would also reduce a bit of complexity and boilerplate brought by type params?

Reply:
I like the type params because it forces the application developer to be up-front about declaring the services and the message formats.  It deliberately means the application developer cannot bury something like this deep in the processing code:

```scala
try {
  // something
} catch {
  case e: RuntimeException =>
    // I haven't really thought about this scenario, so I'll just send an alert
    appHealth.becomeUnhealthyForSetup(Alert("hey pondzix we should fix this"))
}
```

It means there is a single place to look if we want to see all the possible alerts that an app can send.